File: python/knuth/percent.py
"""Percent-format (.py) document model — Python port of src/format/percent.ts,
same semantics, kept honest by round-tripping the same corpus in tests.

Cells open with "# %%" ("#%%" tolerated, marker preserved verbatim);
"# %% [markdown]" is a text cell; "# %% scratch" (exact token) is a scratch
cell; anything else is a program cell. Outputs are machine-managed "#->"
comment lines forming the trailing run of their cell.
"""

import re
from dataclasses import dataclass, field

MARKER = re.compile(r"^# ?%%(.*)$")
OUTPUT_PREFIX = "#->"


@dataclass
class Cell:
    kind: str  # 'program' | 'scratch' | 'text'
    marker: str
    source: list = field(default_factory=list)
    output: list = field(default_factory=list)
    trailing: list = field(default_factory=list)


@dataclass
class Document:
    preamble: list = field(default_factory=list)
    cells: list = field(default_factory=list)
    trailing_newline: bool = True


def _cell_kind(marker_rest):
    rest = marker_rest.strip()
    if rest.startswith("[markdown]"):
        return "text"
    if rest == "scratch":
        return "scratch"
    return "program"


def _split_body(body):
    end = len(body)
    while end > 0 and body[end - 1].strip() == "":
        end -= 1
    start = end
    while start > 0 and body[start - 1].startswith(OUTPUT_PREFIX):
        start -= 1
    if start == end:
        return body, [], []
    return body[:start], body[start:end], body[end:]


def parse_document(text):
    if text == "":
        return Document(trailing_newline=False)
    lines = text.split("\n")
    trailing_newline = lines[-1] == ""
    if trailing_newline:
        lines.pop()

    doc = Document(trailing_newline=trailing_newline)
    current = None  # (marker, rest, body)

    def close():
        if current is None:
            return
        marker, rest, body = current
        source, output, trailing = _split_body(body)
        doc.cells.append(Cell(_cell_kind(rest), marker, source, output, trailing))

    for line in lines:
        m = MARKER.match(line)
        if m:
            close()
            current = (line, m.group(1), [])
        elif current is not None:
            current[2].append(line)
        else:
            doc.preamble.append(line)
    close()
    return doc


def serialize_document(doc):
    lines = list(doc.preamble)
    for c in doc.cells:
        lines.append(c.marker)
        lines.extend(c.source)
        lines.extend(c.output)
        lines.extend(c.trailing)
    if not lines:
        return ""
    return "\n".join(lines) + ("\n" if doc.trailing_newline else "")


def cell_code(cell):
    return "\n".join(cell.source)


def set_output(cell, text):
    """Replace a cell's output block (canonical "#-> " prefixing); None
    clears it. Blank separator lines between cells stay put."""
    if text is None:
        cell.output = []
        cell.source.extend(cell.trailing)
        cell.trailing = []
        return
    if not cell.output:
        end = len(cell.source)
        while end > 0 and cell.source[end - 1].strip() == "":
            end -= 1
        cell.trailing = cell.source[end:]
        cell.source = cell.source[:end]
    cell.output = [
        OUTPUT_PREFIX if line == "" else f"{OUTPUT_PREFIX} {line}" for line in text.split("\n")
    ]

File: src/format/percent.ts
// Percent-format (.py) document model: parse/serialize with byte-identical
// round-tripping. Syntax decisions (recorded in DESIGN.md):
//   - Cells open with "# %%" ("#%%" tolerated, marker preserved verbatim).
//     "# %% [markdown]" is a text cell; "# %% scratch" (exact token) is a
//     scratch cell; any other suffix (titles, tags=...) is a program cell,
//     so jupytext/VS Code/Spyder files parse unchanged.
//   - Outputs are machine-managed comment lines prefixed "#->" forming the
//     trailing run in their cell — no start/end delimiters. Blank output
//     lines are bare "#->". Figures are referenced by path, never embedded.

export type CellKind = 'program' | 'scratch' | 'text';

export interface Cell {
  kind: CellKind;
  /** Raw marker line, preserved verbatim ("# %%", "#%% title", ...). */
  marker: string;
  /** Raw body lines above the output block. */
  source: string[];
  /** Raw "#->" output lines (machine-managed). */
  output: string[];
  /** Raw blank lines between the output block and the next cell. */
  trailing: string[];
}

export interface KnuthDocument {
  /** Raw lines before the first marker (shebang, jupytext header, plain scripts). */
  preamble: string[];
  cells: Cell[];
  /** Whether the file ends with a newline. */
  trailingNewline: boolean;
}

const MARKER = /^# ?%%(.*)$/;
const OUTPUT_PREFIX = '#->';

function cellKind(markerRest: string): CellKind {
  const rest = markerRest.trim();
  if (rest.startsWith('[markdown]')) return 'text';
  if (rest === 'scratch') return 'scratch';
  return 'program';
}

function splitBody(body: string[]): Pick<Cell, 'source' | 'output' | 'trailing'> {
  let end = body.length;
  while (end > 0 && body[end - 1].trim() === '') end--;
  let start = end;
  while (start > 0 && body[start - 1].startsWith(OUTPUT_PREFIX)) start--;
  if (start === end) return { source: body, output: [], trailing: [] };
  return {
    source: body.slice(0, start),
    output: body.slice(start, end),
    trailing: body.slice(end),
  };
}

export function parseDocument(text: string): KnuthDocument {
  if (text === '') return { preamble: [], cells: [], trailingNewline: false };
  const lines = text.split('\n');
  const trailingNewline = lines[lines.length - 1] === '';
  if (trailingNewline) lines.pop();

  const preamble: string[] = [];
  const cells: Cell[] = [];
  let current: { marker: string; rest: string; body: string[] } | null = null;

  const close = () => {
    if (!current) return;
    cells.push({
      kind: cellKind(current.rest),
      marker: current.marker,
      ...splitBody(current.body),
    });
  };

  for (const line of lines) {
    const m = MARKER.exec(line);
    if (m) {
      close();
      current = { marker: line, rest: m[1], body: [] };
    } else if (current) {
      current.body.push(line);
    } else {
      preamble.push(line);
    }
  }
  close();
  return { preamble, cells, trailingNewline };
}

export function serializeDocument(doc: KnuthDocument): string {
  const lines = [
    ...doc.preamble,
    ...doc.cells.flatMap((c) => [c.marker, ...c.source, ...c.output, ...c.trailing]),
  ];
  if (lines.length === 0) return '';
  return lines.join('\n') + (doc.trailingNewline ? '\n' : '');
}

/** Code text of a program/scratch cell. */
export function cellCode(cell: Cell): string {
  return cell.source.join('\n');
}

/** Prose of a text cell, comment prefix stripped. */
export function textProse(cell: Cell): string {
  return cell.source.map((l) => l.replace(/^# ?/, '')).join('\n');
}

/** Replace a text cell's prose (canonical "# " prefixing, blank lines as "#"). */
export function setProse(cell: Cell, prose: string): void {
  cell.source = prose === '' ? [] : prose.split('\n').map((l) => (l === '' ? '#' : `# ${l}`));
}

/** Output block content, "#->" prefix stripped. */
export function outputText(cell: Cell): string {
  return cell.output.map((l) => l.replace(/^#-> ?/, '')).join('\n');
}

/**
 * Replace a cell's output block (canonical "#-> " prefixing); null clears it.
 * Blank separator lines between cells stay put in either direction.
 */
export function setOutput(cell: Cell, text: string | null): void {
  if (text === null) {
    cell.output = [];
    cell.source.push(...cell.trailing);
    cell.trailing = [];
    return;
  }
  if (cell.output.length === 0) {
    let end = cell.source.length;
    while (end > 0 && cell.source[end - 1].trim() === '') end--;
    cell.trailing = cell.source.slice(end);
    cell.source = cell.source.slice(0, end);
  }
  cell.output = text.split('\n').map((l) => (l === '' ? OUTPUT_PREFIX : `${OUTPUT_PREFIX} ${l}`));
}
